modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,764 @@
+		// need to act on it right away (we'll update the set once this header gets finalized), but
+		// we need to make a note of it.
+		//
+		// Note: This assumes that we can only have one authority set change pending at a time.
+		// This is not strictly true as Grandpa may schedule multiple changes on a given chain
+		// if the "next next" change is scheduled after the "delay" period of the "next" change
+		let requires_justification = if let Some(change) = self.storage.scheduled_set_change() {
+			change.height == *header.number()
+		} else {
+			// Since we don't currently have a pending authority set change let's check if the header
+			// contains a log indicating when the next change should be.
+			if let Some(change) = find_scheduled_change(&header) {
+				let next_set = AuthoritySet {
+					authorities: change.next_authorities,
+					set_id: self.storage.current_authority_set().set_id + 1,
+				};
+
+				let height = (*header.number())
+					.checked_add(&change.delay)
+					.ok_or(ImportError::ScheduledHeightOverflow)?;
+				let scheduled_change = ScheduledChange {
+					authority_set: next_set,
+					height,
+				};
+
+				self.storage.schedule_next_set_change(scheduled_change);
+
+				// If the delay is 0 this header will enact the change it signaled
+				height == *header.number()
+			} else {
+				false
+			}
+		};
+
+		let is_finalized = false;
+		self.storage.write_header(&ImportedHeader {
+			header,
+			requires_justification,
+			is_finalized,
+		});
+
+		Ok(())
+	}
+
+	/// Verify that a previously imported header can be finalized with the given Grandpa finality
+	/// proof. If the header enacts an authority set change the change will be applied once the
+	/// header has been finalized.
+	pub fn verify_finality(&mut self, hash: H::Hash, proof: FinalityProof) -> Result<(), FinalizationError> {
+		// Make sure that we've previously imported this header
+		let header = self
+			.storage
+			.header_by_hash(hash)
+			.ok_or(FinalizationError::UnknownHeader)?;
+
+		// We don't want to finalize an ancestor of an already finalized
+		// header, this would be inconsistent
+		let last_finalized = self.storage.best_finalized_header();
+		if header.number() <= last_finalized.number() {
+			return Err(FinalizationError::OldHeader);
+		}
+
+		let current_authority_set = self.storage.current_authority_set();
+		let is_finalized = prove_finality(&header, &current_authority_set, proof);
+		if !is_finalized {
+			return Err(FinalizationError::UnfinalizedHeader);
+		}
+
+		let mut finalized_headers =
+			if let Some(ancestors) = headers_between(&self.storage, last_finalized, header.clone()) {
+				// Since we only try and finalize headers with a height strictly greater
+				// than `best_finalized` if `headers_between` returns Some we must have
+				// at least one element. If we don't something's gone wrong, so best
+				// to die before we write to storage.
+				assert_eq!(ancestors.is_empty(), false);
```

Comment:
Maybe add a description here?
```suggestion
				assert_eq!(ancestors.is_empty(), false, "Emtpy ancestry list returned from headers_between");
```